core/headerchain.go

```diff
@@ -446,7 +446,8 @@ func (hc *HeaderChain) GetTd(hash common.Hash, number uint64) *big.Int {
 	}
 	td := rawdb.ReadTd(hc.chainDb, hash, number)
 	if td == nil {
-		return nil
+		log.Crit("could not find the difficulty for block", "hash", hash)
```

Comment:
A lookup for a non-existent td should be allowed? We shouldn't panic here (`log.Crit`)

And also we can't return `big.Int(0)` here since we have a lot of checks like this in our codebase

```go
var (
		localTD  = f.chain.GetTd(current.Hash(), current.Number.Uint64())
		externTd = f.chain.GetTd(header.Hash(), header.Number.Uint64())
	)
	if localTD == nil || externTd == nil {
		return false, errors.New("missing td")
	}
```

Reply:
fair enough, there are locations in the code that assume a `nil` td is an error, and I could look for all the locations that do not perform this check and capture the code there. I still think it's not a good practice to return `nil` to indicate an error, otherwise we might as well write C code instead of Go (I wouldn't complain about that, mind you).

Reply:
But semantically I think it's correct to return nil in case of a non-existent td. And would be nice to align with some other APIs like `GetHeader` or `GetHeaderByHash`.

Reply:
I get the reasoning but I'm not convinced, because it's still used to convey a number and an error code.